### PR TITLE
Multiple simple improvements (main)

### PR DESCRIPTION
**Pull request by korydraughn**
Still need to write a test for the microservice.

### CODE REVIEW

File: server/re/include/irods/reAction.hpp
Comment (trel) [inline server/re/include/irods/reAction.hpp:67]: i notice these parameters are not consistent with the ones around it... others are all `_t`...  can you explain why?

Comment (korydraughn) [reply server/re/include/irods/reAction.hpp:67]: `MsParam` is the concrete type name while `msParam_t` is a type alias for `MsParam`.

They are identical to the implementation. `msParam_t` exists because of C.

Comment (trel) [reply server/re/include/irods/reAction.hpp:67]: gotcha - so we *could* update all the others if we wanted to 'modernize'...   but not today.

Comment (korydraughn) [reply server/re/include/irods/reAction.hpp:67]: That's correct.

Modernize is probably not the right word for this since `msParam_t` isn't outdated. I just chose to use that today. Could change my mind tomorrow for any number of reasons :-)

File: server/re/CMakeLists.txt
Comment (korydraughn) [inline server/re/CMakeLists.txt:64]: Need to think about this a little more.

Need the user administration library (which relies on a type from the zone administration) for a new microservice (i.e. `msiRemoveUserFromGroup`), hence why I added this.

I'm wonder if this should use `$<BUILD_INTERFACE:...>` and/or whether it should be in the public section.

Is there a better way to do this?

@SwooshyCueb Thoughts?

**Comment by korydraughn** [comment]
Unit tests pass.

Running core tests now.

File: server/re/CMakeLists.txt
Comment (SwooshyCueb) [reply server/re/CMakeLists.txt:64]: Yeah you want `$<BUILD_INTERFACE:...>` and in the `PUBLIC` section.

Comment (korydraughn) [reply server/re/CMakeLists.txt:64]: Can you explain how you arrived at that conclusion?

**Comment by korydraughn** [comment]
All core tests passed.

Will update the CMake and re-run the tests to be on the safe side.

**Comment by korydraughn** [comment]
Running core tests with cmake updates.

**Comment by korydraughn** [comment]
Core tests passed.

File: server/re/include/irods/reAction.hpp
Comment (korydraughn) [reply server/re/include/irods/reAction.hpp:67]: Do you have a preference for one over the other?

Comment (trel) [reply server/re/include/irods/reAction.hpp:67]: no, defer to you and team.   just caught my eye is all.

**Review by alanking — COMMENTED** [review]
We need tests for the new microservice, yes?

File: server/re/src/icatAdminMS.cpp
Comment (korydraughn) [reply server/re/src/icatAdminMS.cpp:360]: I had a similar thought. The fact that you also mentioned it means it's probably a good idea to do it.

Will change.

**Comment by korydraughn** [comment]
> We need tests for the new microservice, yes?

Yes. Will add tests for bad inputs.

**Comment by alanking** [comment]
Oh, just saw the tests. Disregard last comment

**Comment by korydraughn** [comment]
Well, we cover good inputs, but not bad ones. We definitely need to cover the negative space.

File: server/re/CMakeLists.txt
Comment (korydraughn) [reply server/re/CMakeLists.txt]: Created irods/irods_docs#208 to capture CMake knowledge in the developer docs.

File: server/re/include/irods/reAction.hpp
Comment (korydraughn) [reply server/re/include/irods/reAction.hpp:67]: Leaving as is. This isn't urgent at the moment.

**Comment by korydraughn** [comment]
Added python tests for bad inputs and remote users.

The new tests pass.
Re-ran the unit tests and they passed too.

Please review and let me know if anything looks suspicious.

**Review by alanking — COMMENTED** [review]
Seems good to me. Squash as desired and leave off #s

**Comment by trel** [comment]
didn't see a test that shows it works / doesn't break when the requested user to be removed is NOT part of that group.

**Comment by korydraughn** [comment]
Ah, will add a test.

**Comment by korydraughn** [comment]
Now the question becomes ... should the MSI return an error when attempting to remove the user from the group, but the user isn't a member of it?

At the moment, `CAT_SUCCESS_BUT_WITH_NO_INFO` is returned which **could** mean the user isn't a member, but no one can rely on that exactly.

Should anyone care about this case? I don't think there's much value in knowing whether the removal failed in this case. Why would someone want to know that? They can always query for that info.

Thoughts?

**Comment by korydraughn** [comment]
New test passed at the bench which means all tests pass.

**Comment by korydraughn** [comment]
Ubuntu returns an error if the user isn't a member of the group.
```bash
sudo addgroup testgroup
sudo deluser <username> testgroup
/usr/sbin/deluser: The user `<username>` is not a member of group `testgroup`.
```

However, does it make sense to return an error in a policy context?

Should we add an option that changes the MSI's behavior?

An additional option would move the decision to the caller. Thoughts?

**Comment by trel** [comment]
unix behavior is a strong indicator.    i think we need a reason to diverge from that answer.

`CAT_SUCCESS_BUT_WITH_NO_INFO` is a poor return code... do we have something else that complains more correctly?

**Comment by korydraughn** [comment]
Not sure we have an error that captures that case. Can add one though.

How about `USER_NOT_A_MEMBER_OF_GROUP`?

That code should likely come from the database plugin.

**Comment by trel** [comment]
so it's historically always given the `CAT_SUCCESS_BUT_WITH_NO_INFO` error?

**Comment by korydraughn** [comment]
That's my guess.

The MSI calls the user administration library, which calls rsGeneralAdmin. Everything's a thin wrapper around rsGeneralAdmin.

**Comment by trel** [comment]
Then let’s just leave it for now.   It’s not broken, just a bit esoteric.

**Comment by korydraughn** [comment]
Sure. I'll write up an issue to introduce a better error code for this case.

Can handle it in a later PR.

**Comment by korydraughn** [comment]
Created new issue: #7216

**Comment by korydraughn** [comment]
With that, I'm going to squash the changes.

**Comment by korydraughn** [comment]
Squashed.

Checking the issue numbers for correctness.

**Comment by SwooshyCueb** [comment]
There's a lot of `clang-tidy` whinging. I'd like to see the `bugprone-lambda-function-name` check failures in `icatAdminMS.cpp` either addressed or disabled with `NOLINTNEXTLINE`, but I don't feel that strongly about it if others feel otherwise

**Comment by korydraughn** [comment]
Good catch. Will fix.

**Comment by SwooshyCueb** [comment]
Looks like we still have a `bugprone-lambda-function-name` check failure
https://github.com/irods/irods/actions/runs/5683534551/job/15404295230?pr=7209#step:7:58

**Comment by korydraughn** [comment]
I'm pretty sure that one is not real. The HTTP API does the same thing and results in the correct behavior. I'll verify with GCC to see if it results in different behavior. I'm hoping it handles the `fn = __func__` like Clang. If not, then I'll pull it out.

Will take care of all the workflow results.

**Comment by korydraughn** [comment]
I've confirmed GCC handles the capture of the function name the same way as Clang. Address sanitizer doesn't report any issues either, so we're good.

Will add a `NOLINT` directive to the lambda.

**Review by alanking — COMMENTED** [review]
Looks good to me

**Review by SwooshyCueb — APPROVED** [review]

**Comment by korydraughn** [comment]
I made a change to one of the serialization routines where clang-tidy reported a performance issue.

Will run through the test suite again to confirm everything is still good.

**Comment by korydraughn** [comment]
All unit tests passed.

Running core tests.

File: server/re/src/irods_re_serialization.cpp
Comment (korydraughn) [inline server/re/src/irods_re_serialization.cpp:986]: Capture the bugprone clang-tidy error in an issue.

**Comment by korydraughn** [comment]
All core tests passed.

Squashing now.

**Comment by korydraughn** [comment]
I changed the formatting for a line that was really long, but left the actual logic as it was.

Will add the pounds once approved.

**Comment by korydraughn** [comment]
Just to be safe, I ran the unit tests and they all passed.

File: server/re/src/irods_re_serialization.cpp
Comment (alanking) [reply server/re/src/irods_re_serialization.cpp]: Is this done?

**Review by alanking — APPROVED** [review]
Yep, add #s when ready.

File: server/re/src/irods_re_serialization.cpp
Comment (korydraughn) [reply server/re/src/irods_re_serialization.cpp]: The line in question will change with this PR. I'm going to create the issue after this is merged.

Comment (alanking) [reply server/re/src/irods_re_serialization.cpp]: Makes sense. So this will be resolved after merging, then.

Comment (korydraughn) [reply server/re/src/irods_re_serialization.cpp]: Yep

**Comment by korydraughn** [comment]
Added #'s

File: server/re/src/irods_re_serialization.cpp
Comment (korydraughn) [reply server/re/src/irods_re_serialization.cpp]: Created #7222

**Pull request merged**
